### DISCUSSION TITLE
3.2 enough said.

### DISCUSSION THREAD

**Discussion by jdonwells** [TEALSK12/introduction-to-computer-science#344] [Q&A] [answered]
Just looking at 3.2.

In SUMMARY.md it says "Do Now 3.2: Drawing Squares". I don't think that title is correct. The document says "Sprite Control".

Lab 3.2. The instructions are not clear.

2.2 In 2.1 we used the green flag. How is 2.1 going to be graded for the rubric if the students disassemble it to make part 2.2 work? Should the dino move to the top and then to the bottom? Should it move to the top and stop till the green flag is clicked a second time? Should it do both at the same time and just stand still? If I don't know then the students surely don't know.

paragraph after 2.3. When was touching color introduced? If you want them to use it shouldn't it be mentioned in the slide show?

3.1 Should the Dino move to the top of the stage and then move to a random location? Should it move diagonally both upwards and to one side or the other? Should it move straight to the top then jump to a random x or should it be animated?

3.2 Does the Dino fall? When does the Dino fall? Is this from 2.2? 

paragraph after 4.1. There is no change in y block there is only a change y by block. What forever block is this still 2.2?

The starter code offers us two green flag clicked and forever blocks. What are the students supposed to do with that? Is this supposed to clarify the instructions for 2.2? I see it as confusing to 2.1.

If you are simulating gravity as stated by the slide show: "construct a simple model for gravity." Then why not actually do so by adding a fixed acceleration to the speed?

**Comment by jdonwells** [comment, marked as answer]
I went ahead and added slides to the slide show for simulating gravity. https://docs.google.com/presentation/d/18MD-Gf_N155OSFzM9CI2vPdHDiabUYin30aUm_Umm8E/edit?usp=sharing

I changed the lab write up to this:

# Lab 3.2 - What Goes Up

In this lab, you will use everything you've learned about loops and conditionals to construct a simple model for gravity.

## Part 1: Before You Start

Go to the starter project at [https://aka.ms/Intro-lab3-2](https://aka.ms/Intro-lab3-2). Log into Snap! and save your own copy of the project by choosing `Save as` button from the file menu. Be sure to click the `Share` button in the Save dialog box. Highlight the URL in the address bar and copy it.

## Part 2: Channeling Newton

Write a script for the Dino sprite so that:

2.1) Use one `when green flag is clicked`, to place Dino at the top of the stage. No animation, just `set y to`.

2.2) Use two more `when green flag is clicked`, to start two `forever` loops that make Dino fall to the bottom of the stage. One will model gravity by subtracting a constant amount from Dino's speed. The second will move Dino by his current speed.

2.3) When Dino touches the ground, Dino stops falling. "*Stops falling*" means Dino's speed is set to zero and stays zero as long as he is touching the ground. The `touching color ?` block can be used to detect the grass in the background layer.

When using the `touching color` block, change the color by clicking on the color box, then select anything in the color palette or stage that has the desired color.

## Part 3: What goes up, Must Come Down

3.1) Modify your code so that when the `green flag` is clicked, Dino will still move to the top of the stage but also move to a random x-coordinate between -240 and 240.  Use the `pick random to` block in the `operators` category.

3.2) Modify your code so that the Dino sprite stops falling when it touches the ground `or` if it is touching the Platform.

## Part 4: Jump Up

4.1) Modify your code so that when the `spacebar` is pressed, Dino will jump up. From a physics point of view, when you jump you set your speed to something greater than zero.

## Part 5: Challenge

5.1) Modify your code so Dino can only jump while on the ground or the platform.

5.2) Add code so Dino will move left and right with the arrow keys.

## Grading Scheme/Rubric

| **Lab 3.2 Criteria** |   Points |
| --- | --- |
| 2.1 Goes to the top of the stage | 0.3 |
| 2.2 Falls to the bottom of the stage | 0.3     |
| 2.3 Stops falling when touching the ground | 0.3 |
| 3.1 To a random x-coordinate | 0.2 |
| 3.2 Touching the ground OR touching the platform  | 0.3  |
| 4.1 Jump up | 0.3 |
| 5.1 Jump Up: Ground or Platform | 0.5 |
| 5.2 Movement using left and right arrow  | 0.2  |
| **PROJECT TOTAL** | **2.5 points** |

Here is my solution to the above lab.
[https://snap.berkeley.edu/snap/snap.html#present:Username=jdonwells&ProjectName=Lab%203.2%20Dino](https://snap.berkeley.edu/snap/snap.html#present:Username=jdonwells&ProjectName=Lab%203.2%20Dino)

**Reply by jpfasano** [reply]
@jdonwells Thank you for your solution.  I learned some new techniques from viewing it.

With your solution if Dino moves below the platform and jumps, then Dino sticks to the platform and does not fall back down to the grass.
One can continue to press the space bar and Dino will jump up through the platform.
I don't think this is the expected behavior.
At the moment I'm not sure of a good fix.

**Reply by jdonwells** [reply]
To fix that, you need to add the idea that Dino can touch the platform by jumping up or falling down. To me, that is a separate lab. But the question you have to ask as a teacher is does it add value? To make the perfect platform game it does. To learn about abstraction and generalization it doesn't. My experience is that students don't get that far in the unit project to worry about those kinds of details.

**Reply by jpfasano** [reply]
@jdonwells  Yes, I completely agree that pedagogically it does not make sense to expect the students write a snap code to fix this issue.  However, it does make sense to write the assignment to match the code we want the students to write.  I've submitted a git pull request that updates the lab's wording.

**Comment by aspiece** [comment]
@jdonwells  This is great. Thank you. I created an issue so we can get this into our next PR.
